### ISSUE TITLE
Add specialized accessor operations to `Chunk`

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#3386]
As the first step toward having specialized chunks for different primitive types, we can add accessors that are also specialized, which can avoid the overhead of boxing for single element access:

```scala
sealed trait Chunk[+A] {
  def boolean(index: Long)(implicit ev: A <:< Boolean): Boolean = ev(apply(index))
  def byte(index: Long)(implicit ev: A <:< Byte): Byte = ev(apply(index))
  def char(index: Long)(implicit ev: A <:< Char): Char = ev(apply(index))
  def short(index: Long)(implicit ev: A <:< Short): Short = ev(apply(index))
  def int(index: Long)(implicit ev: A <:< Int): Int = ev(apply(index))
  def long(index: Long)(implicit ev: A <:< Long): Long = ev(apply(index))
  def float(index: Long)(implicit ev: A <:< Float): Float = ev(apply(index))
  def double(index: Long)(implicit ev: A <:< Double): Double = ev(apply(index))
}
```

These default implementations will be slow, but that's not the point: they can be overridden in specialized children to perform no boxing.

This ticket merely involves adding the above accessors to `Chunk`, alphabetically, with a few tests, perhaps, but **not** making additional changes necessary for efficient implementations in specialized chunks.

**Comment by whysoserious** [comment]
I'll take it.

**Comment by adamgfraser** [comment]
Great!

**Comment by whysoserious** [comment]
It looked easy but I stumbled upon a new bug. New tests added in PR https://github.com/zio/zio/pull/3431 time out. 
Example test which times out:
```scala
      check(chunkWithIndex(Gen.short(5, 100))) {
        case (chunk, i) =>
          assert(chunk.short(i))(equalTo(chunk.toList.apply(i)))
      }
```
Function `chunkWithIndex` does not generate required number of `Chunk`s (200), unit test freezes waiting for more input and times out after 1 minute.
However `chunkWithIndex(Gen.unit)` works as expected.

**Comment by whysoserious** [comment]
Uff, got it.

**Comment by adamgfraser** [comment]
Nice! Was there a circularity?

**Comment by whysoserious** [comment]
Yes, `ChunkSpec.chunkWithIndex` was hanging while trying to generate a random index in an empty Chunk. PR is coming soon.

**Comment by whysoserious** [comment]
@adamgfraser #3431

**Comment by adamgfraser** [comment]
Resolved by #3431.
